--- a/src/mod/endpoints/mod_rtmp/handshake.h
+++ b/src/mod/endpoints/mod_rtmp/handshake.h
@@ -35,6 +35,12 @@
 #include <openssl/hmac.h>
 #include <openssl/sha.h>
 
+#ifdef LIBRESSL_VERSION_NUMBER
+#include <openssl/rand.h>
+#undef OPENSSL_VERSION_NUMBER
+#define OPENSSL_VERSION_NUMBER 0x1000107fL
+#endif
+
 #if !defined(_MSC_VER) && !defined(__FUNCTION__)
 #define __FUNCTION__ (const char *)__func__
 #endif

