--- a/src/mod/event_handlers/mod_kazoo/Makefile.am
+++ b/src/mod/event_handlers/mod_kazoo/Makefile.am
@@ -14,10 +14,12 @@
 mod_kazoo_la_LIBADD   = $(switch_builddir)/libfreeswitch.la
 mod_kazoo_la_LDFLAGS  = -avoid-version -module -no-undefined -shared @ERLANG_LDFLAGS@
 
+mod_kazoo.la: kazoo_definitions.h $(mod_kazoo_la_OBJECTS) $(mod_kazoo_la_DEPENDENCIES) $(EXTRA_mod_kazoo_la_DEPENDENCIES)
+	$(AM_V_CCLD)$(mod_kazoo_la_LINK) $(am_mod_kazoo_la_rpath) $(mod_kazoo_la_OBJECTS) $(mod_kazoo_la_LIBADD) $(LIBS)
+
+
 kazoo_definitions.h:
 	xxd -i kazoo.conf.xml > kazoo_definitions.h
-
-local_depend: kazoo_definitions.h
 
 else
 install: error
